packages/kbn-i18n/src/core/i18n.js

```diff
@@ -0,0 +1,240 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/**
+ @typedef Messages - messages tree, where leafs are translated strings
+ @type {object<string, object>}
+ @property {string} [locale] - locale of the messages
+ @property {object} [formats] - set of options to the underlying formatter
+ */
+
+import IntlMessageFormat from 'intl-messageformat';
+import IntlRelativeFormat from 'intl-relativeformat';
+import memoizeIntlConstructor from 'intl-format-cache';
+import { isString, isObject, hasValues, deepMerge } from './helper';
+import { formats as EN_FORMATS } from './formats';
+
+// Add all locale data to `IntlMessageFormat`.
+// TODO: Use dynamic import for asynchronous loading of specific locale data
+import './locales';
+
+const EN_LOCALE = 'en';
+const LOCALE_DELIMITER = '-';
+const messages = {};
+const getMessageFormat = memoizeIntlConstructor(IntlMessageFormat);
+
+let defaultLocale = EN_LOCALE;
+let currentLocale = EN_LOCALE;
+let formats = EN_FORMATS;
+
+IntlMessageFormat.defaultLocale = defaultLocale;
+IntlRelativeFormat.defaultLocale = defaultLocale;
+
+/**
+ * Returns message by the given message id.
+ * @param {Messages} messages - messages tree
+ * @param {string} id - path to the message that consists of properties
+ * names separated by dots
+ * @returns {string} message - translated message from messages tree
+ * @example
+ * getMessageById({ a: { b: { c: 'test' } } }, 'a.b.c'); // => 'test'
```

Comment:
I've been a little out of the loop here: do we know that nested translations will be useful? I'm not against it, but I had only really considered having a flat json shape with basic key/value pairs.

Reply:
From my perspective, it's easier to navigate through namespaces in nested-JSONs and we don't need to copy-paste common part of ids for several messages under same namespace.

Reply:
I'm not super confident about that either, but gave it some thought and still leaning towards flat JSON key-value pairs since nobody except for localizers is going to look into these files so it'd probably make sense to optimize for them and we load it once, cache and then call `translate` many times so in the end splitting on the fly may cost us more than larger translations file (because of same namespace for many labels).

Here is feedback from Chi-Wei, just for the record:

```
Plain key-value map (full path) is preferable because it quickly provides full context at a glance
for where a string is coming from, both helpful for translators and for speeding up troubleshooting.

Nested file paths may also create the potential for more read failures if a referenced property
 hasn’t been defined or localized yet.

As with most things in localization though, vendors and linguists are used to working with both
 nested and plain key formats, so if it will require a lot of work to reformat, it’s not the end of 
the world to use the nested structure.
```

What do you think @maksim-tolo, @archanid @epixa ?

Reply:
++ flat